File: ruby/sample.rb
require 'sinatra'
require 'openssl'
require 'Base64'
require 'JSON'

set :port, 8080


################################################
#            Config section start              #
#         replace with your own values         #
################################################

SECRET_KEY = "YourSecretKey"
CREDENTIAL_TTL = 3600

#################################################
#            Config section finished            #
#################################################

post '/authenticate' do
  headers \
      "Access-Control-Allow-Origin" => "*",
      "Content-Type" => "application/json"

  peer_id = params['peerId']
  session_token = params['sessionToken']

  if peer_id == nil || session_token == nil
    status 400
    return
  end

  if check_session_token(peer_id, session_token)
    # Session token check was successful.
    unix_timestamp = Time.now.to_i

    credential = {
        peerId: peer_id,
        timestamp: unix_timestamp,
        ttl: CREDENTIAL_TTL,
        authToken: calculate_auth_token(peer_id, unix_timestamp)
    }

    body JSON.generate(credential)
  else
    # Session token check failed.
    status 401
  end
end

def check_session_token(peer_id, token)
  # Implement checking whether the session is valid or not.
  # Return true if the session token is valid.
  # Return false if it is invalid.
  true
end

def calculate_auth_token(peer_id, timestamp)
  message = "#{timestamp}:#{CREDENTIAL_TTL}:#{peer_id}"
  hash = OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha256'), SECRET_KEY, message)
  Base64.encode64(hash).strip()
end

File: nodejs/sample.js
'use strict';

const express = require('express');
const bodyParser = require('body-parser');

const hmac = require('crypto-js/hmac-sha256');
const CryptoJS = require('crypto-js');

/************************************************
 *            Config section start              *
 *         replace with your own values         *
 ************************************************/

const secretKey = 'YourSecretKey'; // replace with your own secretKey from the dashboard
const credentialTTL = 3600; // 1 hour

/************************************************
 *            Config section finished           *
 ************************************************/

const app = express();
app.use(bodyParser.json()); // support json encoded bodies
app.use(bodyParser.urlencoded({ extended: true })); // support encoded bodies
app.use(function(req, res, next) {
  res.header("Access-Control-Allow-Origin", "*");
  res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
  next();
});

app.post('/authenticate', (req, res) => {
  const peerId = req.body.peerId;
  const sessionToken = req.body.sessionToken;

  if(peerId === undefined || sessionToken === undefined) {
    res.status(400).send('Bad Request');
    return;
  }

  checkSessionToken(peerId, sessionToken).then(() => {
    // Session token check was successful.

    // We need the current unix timestamp. Date.now() returns in milliseconds so divide by 1000 to get seconds.
    const unixTimestamp = Math.floor(Date.now() / 1000);

    const credential = {
      peerId: peerId,
      timestamp: unixTimestamp,
      ttl: credentialTTL,
      authToken: calculateAuthToken(peerId, unixTimestamp)
    };

    res.send(credential);
  }).catch(() => {
    // Session token check failed
    res.status(401).send('Authentication Failed');
  });
});

const listener = app.listen(process.env.PORT || 8080, () => {
  console.log(`Server listening on port ${listener.address().port}`)
});

function checkSessionToken(peerId, token) {
  return new Promise((resolve, reject) => {
    // Implement checking whether the session is valid or not.
    // Resolve if the session token is valid.
    // Reject if it is invalid.

    resolve();
  });
}

function calculateAuthToken(peerId, timestamp) {
  // calculate the auth token hash
  const hash = CryptoJS.HmacSHA256(`${timestamp}:${credentialTTL}:${peerId}`, secretKey);

  // convert the hash to a base64 string
  return CryptoJS.enc.Base64.stringify(hash);
}